src/table_properties.rs

```diff
@@ -0,0 +1,248 @@
+    fn add_user_key(
+        &mut self,
+        key: &[u8],
+        value: &[u8],
+        entry_type: EntryType,
+        seq: u64,
+        file_size: u64,
+    ) -> bool;
+
+    /// Called after each new block is cut
+    fn block_add(
+        &mut self,
+        _block_uncompressed_bytes: u64,
+        _block_compressed_bytes_fast: u64,
+        _block_compressed_bytes_slow: u64,
+    ) {
+    }
+
+    /// Called once when a table has been built and is ready for writing the properties block
+    ///
+    /// When the returned Status is false, the collected properties will not be written to the
+    /// file's property block.
+    fn finish(&mut self, properties: &mut HashMap<String, String>) -> bool;
+
+    /// Returns human-readable properties used for logging
+    ///
+    /// It will only be called after finish() has been called.
+    fn get_readable_properties(&self) -> HashMap<String, String>;
+
+    /// Name of the collector to use for logging
+    fn name(&self) -> &CStr;
+}
+
+struct TablePropertiesCollectorFactoryCallback<F>
+where
+    F: TablePropertiesCollectorFactory + Send + 'static,
+{
+    _phantom: PhantomData<F>,
+}
+
+impl<F> TablePropertiesCollectorFactoryCallback<F>
+where
+    F: TablePropertiesCollectorFactory + Send + 'static,
+{
+    unsafe extern "C" fn create_collector(
+        raw_self: *mut c_void,
+        ctx: *mut ffi::rocksdb_table_properties_collector_context_t,
+    ) -> *mut ffi::rocksdb_table_properties_collector_t {
+        let context = TablePropertiesCollectorContext {
+                column_family_id: ffi::rocksdb_table_properties_collector_context_get_column_family_id(ctx),
+                level_at_creation: ffi::rocksdb_table_properties_collector_context_get_level_at_creation(ctx),
+                num_levels: ffi::rocksdb_table_properties_collector_context_get_num_levels(ctx),
+                last_level_inclusive_max_seqno_threshold:
+                    ffi::rocksdb_table_properties_collector_context_get_last_level_inclusive_max_seqno_threshold(ctx),
+            };
+
+        let factory: &mut F = &mut *(raw_self.cast());
+        let collector = Box::new(factory.create(context));
+
+        ffi::rocksdb_table_properties_collector_create(
+            Box::into_raw(collector).cast(),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::destructor),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::add_user_key),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::block_add),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::finish),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::get_readable_properties),
+            Some(TablePropertiesCollectorCallback::<F::Collector>::name),
+        )
+    }
+
+    unsafe extern "C" fn name(raw_self: *mut c_void) -> *const c_char {
+        let factory = &*(raw_self.cast_const() as *const F);
+        factory.name().as_ptr()
+    }
```

Comment:
I didn't trace the entire flow, but can you verify that that we are not leaking strings somewhere, I remember seeing `strdup` somewhere in the C binding re. to names (not sure which) and I just want to make sure that this allocated string is eventually owned and deallocated.